src/analysis/souffle/check_predicate.dl

```diff
@@ -0,0 +1,82 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+#define SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+
+#include "authorization_logic.dl"
+#include "dataflow_graph.dl"
+#include "tags.dl"
+
+.type Bool <: number
+
```

Comment:
Do we have a good sense of what the performance of ADT in datalog is?

Reply:
That is an interesting question for which I cannot find a full answer online. But I can find some clues.

See this commit to Souffle: https://github.com/souffle-lang/souffle/commit/ddca2fdaeef03a3f6de77abff3a86403fe649b58 .

This commit, in its own words, optimizes ADTs by:

```
encode branch with a single argument as [branch_id, argument]
instead of [branch_id, [argument]]
```

What this tells me is that the various fields of a branch of an ADT are represented as a linked list internally. This means that exploding them and placing each field in a relation, as we do to evaluate the ADT, is probably very cheap. What might not be cheap is getting the last field on a branch of an ADT with many fields, as traversing to each successive field incurs an additional pointer dereference. None of our branches, however, have many fields, so this is probably ok.

That's all I know about ADTs in general. However, I think I have a pretty good understanding of how expensive using ADTs for this particular purpose will be. The evaluation of `check` ADTs will occur at the very end of the analysis; it must, because the leaf `TagPresence` branch consults `!mayHaveTag`, introducing a stratifiability requirement. This means that evaluation of the `CheckPredicate` ADTs will not be interleaved with pushing around tags for the analysis. I believe that this means that this will not occur as part of the inner loop of the analysis, and thus will be of least concern from a performance perspective.